### ISSUE TITLE
 sl-app BETA webserver does not start before a random time

### ISSUE THREAD

**Issue by Metabaron1** [simple-login/app#1618]
Hi

I'm testing upgrade from 3.4.0 to 4.6.5-beta and followed
https://github.com/simple-login/app/blob/new-self-host-version/docs/upgrade.md
I run sl-migration(alembic upgrade head) + sl-init + update email_log set alias_id=(select alias_id from contact where contact.id = email_log.contact_id);
But seems that sl-app can't use DB.
Any idea? Or how can I enable debug  on sl-app please?

From sl-app container:
# cat /code/.env|grep ^DB_URI   
DB_URI=postgresql://sladmin:mypassword@sl-db:5432/simplelogin
# host sl-db
sl-db has address 172.18.0.2
# telnet sl-db 5432
Trying 172.18.0.2...
Connected to sl-db.
Escape character is '^]'.
exit
Connection closed by foreign host.
# python shell.py
>>> URL: https://mydomain.fr:4443
Paddle param not set
WARNING: Use a temp directory for GNUPGHOME /tmp/jqxggaslpynbbtihbfcs
Upload files to local dir
>>> init logging <<<
2023-03-05 19:37:04,828 - SL - DEBUG - 2211 - "/code/app/utils.py:14" - <module>() -  - load words file: /code/local_data/words_alpha.txt
Python 3.10.9 (main, Dec 21 2022, 10:27:53) [GCC 10.2.1 20210110]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.31.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: print(User.query.first())
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/code/shell.py in <module>
----> 1 print(User.query.first())

AttributeError: 'function' object has no attribute 'first'

sl-app logs:

Mar  5 16:35:25 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:25 +0100] [1] [INFO] Starting gunicorn 20.0.4
Mar  5 16:35:25 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:25 +0100] [1] [INFO] Listening at: http://0.0.0.0:7777 (1)
Mar  5 16:35:25 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:25 +0100] [1] [INFO] Using worker: sync
Mar  5 16:35:25 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:25 +0100] [7] [INFO] Booting worker with pid: 7
Mar  5 16:35:25 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:25 +0100] [8] [INFO] Booting worker with pid: 8
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: >>> URL:>>> URL: https://mydomain.fr:4443
Mar  5 16:35:29 172.18.0.1 sl-app[28175]:  Paddle param not sethttps://mydomain.fr:4443
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Paddle param not set
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: WARNING: Use a temp directory for GNUPGHOME
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: WARNING: Use a temp directory for GNUPGHOME /tmp/rfdosltmcowkdvvknkgi
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Upload files to local dir
Mar  5 16:35:29 172.18.0.1 sl-app[28175]:  /tmp/punqzyoaqqnteitakjkk
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Upload files to local dir
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: >>> init logging <<<>>> init logging <<<
Mar  5 16:35:29 1Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'mailbox_or_url_list' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'mailbox_or_url_list' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'delim' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'mailbox_or_url' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'url' is unreachable
...
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'mailbox_or_url_list' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'delim' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'mailbox_or_url_list' is unreachable
Mar  5 16:35:29 172.18.0.1 sl-app[28175]: Symbol 'delim' is unreachable
Mar  5 16:35:32 172.18.0.1 sl-app[28175]: 2023-03-05 16:35:32,250 - SL - DEBUG - 7 - "/code/app/utils.py:14" - <module>() -  - load words file: /code/local_data/words_alpha.txt
Mar  5 16:35:32 172.18.0.1 sl-app[28175]: 2023-03-05 16:35:32,252 - SL - DEBUG - 8 - "/code/app/utils.py:14" - <module>() -  - load words file: /code/local_data/words_alpha.txt
Mar  5 16:35:41 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:41 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:7)
Mar  5 16:35:41 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:41 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:8)
Mar  5 16:35:42 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:42 +0100] [17] [INFO] Booting worker with pid: 17
Mar  5 16:35:42 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:42 +0100] [18] [INFO] Booting worker with pid: 18
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: >>> URL: https://mydomain.fr:4443
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: Paddle param not set
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: WARNING: Use a temp directory for GNUPGHOME /tmp/dmjtnyoaddebrragcvsk
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: Upload files to local dir
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: >>> URL: https://mydomain.fr:4443
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: Paddle param not set
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: WARNING: Use a temp directory for GNUPGHOME /tmp/ibnxaqludzttijnmhmap
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: Upload files to local dir
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: >>> init logging <<<
Mar  5 16:35:44 172.18.0.1 sl-app[28175]: >>> init logging <<<
Mar  5 16:35:47 172.18.0.1 sl-app[28175]: 2023-03-05 16:35:47,962 - SL - DEBUG - 17 - "/code/app/utils.py:14" - <module>() -  - load words file: /code/local_data/words_alpha.txt
Mar  5 16:35:48 172.18.0.1 sl-app[28175]: 2023-03-05 16:35:48,022 - SL - DEBUG - 18 - "/code/app/utils.py:14" - <module>() -  - load words file: /code/local_data/words_alpha.txt
Mar  5 16:35:57 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:57 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:17)
Mar  5 16:35:57 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:57 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:18)
Mar  5 16:35:57 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:57 +0100] [17] [INFO] Worker exiting (pid: 17)
Mar  5 16:35:58 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:58 +0100] [18] [INFO] Worker exiting (pid: 18)
Mar  5 16:35:58 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:58 +0100] [27] [INFO] Booting worker with pid: 27
Mar  5 16:35:58 172.18.0.1 sl-app[28175]: [2023-03-05 17:35:58 +0100] [28] [INFO] Booting worker with pid: 28
the 2 workers keeps rebooting...

**Comment by nguyenkims** [comment]
@Metabaron1 in the shell.py, can you try to run `print(User.first())` instead of `print(User.query.first())` instead?

**Comment by Metabaron1** [comment]
@nguyenkims 

Oh, it's OK lol
In [1]: print(User.first())
<User 1  my email>

As I thought it was not working I did not try to go further and stopped upgrade.
But is it a normal behavior that the 2 workers keep rebooting in sl-app logs?

**Comment by Metabaron1** [comment]
No, there is really a probleme with sl-app container, the webserver does not reply from inside the container:

# wget localhost:7777
--2023-03-07 17:21:02--  http://localhost:7777/
Resolving localhost (localhost)... 127.0.0.1, ::1
Connecting to localhost (localhost)|127.0.0.1|:7777... connected.
HTTP request sent, awaiting response... 

I also add +1G (previouly 2G +1G=3G)  RAM as you said in a previous issue that this new webwerver migth use more memory

But the 2 workers keeps rebooting in sl-app logs:

Mar  7 17:37:40 172.22.0.1 sl-app[1484]: [2023-03-07 17:37:40 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:65)
Mar  7 17:37:40 172.22.0.1 sl-app[1484]: [2023-03-07 17:37:40 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:66)
Mar  7 17:37:40 172.22.0.1 sl-app[1484]: [2023-03-07 17:37:40 +0100] [65] [INFO] Worker exiting (pid: 65)
Mar  7 17:37:40 172.22.0.1 sl-app[1484]: [2023-03-07 17:37:40 +0100] [66] [INFO] Worker exiting (pid: 66)
Mar  7 17:37:41 172.22.0.1 sl-app[1484]: [2023-03-07 17:37:41 +0100] [75] [INFO] Booting worker with pid: 75
Mar  7 17:37:41 172.22.0.1 sl-app[1484]: [2023-03-07 17:37:41 +0100] [76] [INFO] Booting worker with pid: 76
...
ar  7 17:38:12 172.22.0.1 sl-app[1484]: [2023-03-07 17:38:12 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:85)
Mar  7 17:38:12 172.22.0.1 sl-app[1484]: [2023-03-07 17:38:12 +0100] [1] [CRITICAL] WORKER TIMEOUT (pid:86)
Mar  7 17:38:12 172.22.0.1 sl-app[1484]: [2023-03-07 17:38:12 +0100] [85] [INFO] Worker exiting (pid: 85)
Mar  7 17:38:13 172.22.0.1 sl-app[1484]: [2023-03-07 17:38:13 +0100] [86] [INFO] Worker exiting (pid: 86)
Mar  7 17:38:13 172.22.0.1 sl-app[1484]: [2023-03-07 17:38:13 +0100] [95] [INFO] Booting worker with pid: 95
Mar  7 17:38:13 172.22.0.1 sl-app[1484]: [2023-03-07 17:38:13 +0100] [96] [INFO] Booting worker with pid: 96

**Comment by Slevin90** [comment]
I remember seeing similar things a few weeks back on my self-hosted setup.
For me it worked to stop and start the containers in the a fixed order.

stop all containers
start them in following order:   (names may have been set different)

1. sl-db
2. sl-app
3. sl-email
4. sl-jobrunner

**Comment by Metabaron1** [comment]
I'm running with docker-compose  with depends-on database. I wasn't aware about such dependency between sl-apps.. but I tried with this order without success. 
I tried from scratch with empty DB like a fresh install,  I even tried with posgresql:13 empty DB.. always same rebooting sl-app worker. Then as it's not working with an empty database, either there is a major change in simplelogin.env  I don't see,, either there is a problem with postfix...

I have updated postfix config as stated in "After upgrade to 4.x.x from 3.4.0" . I know I'm using a bit old (2 years) own postfix image and I've seen there are some changes btree/hash dictionary licence, but my postfix container is still working fine, then this can't prevent sl-app from running correctly?

**Comment by Metabaron1** [comment]
when checking today, the sl-app I startded yesterday and forgot to stop was  responding. When looking log, the workers stopped rebooting after 25min.
Then I tried to stop everything and restart DB and sl-app only and wait... and indeed the sl-app webserver start reponding after a random period of time: half an hour, 5min ...
Not that it's something I use to restart, but that's a really strange behavior.

**Comment by nguyenkims** [comment]
@Metabaron1 do you know if the container memory is somehow limited in your docker-compose? Can you also try with running `python server.py` directly which should consume less memory than the default `gunicorn`

**Comment by Metabaron1** [comment]
I got it! :)

No containers use by defaut all  the RAM, unless you limit it. And sl-app only use 300M, and all others  containers 1G, then 2G RAM if far than enought for self hosting.

But I noticed that the period betwen 2 workrs restart is 15s, which is the timeout you setup in Dockerfile for gunicorn:
CMD ["gunicorn","wsgi:app","-b","0.0.0.0:7777","-w","2","--timeout","15"]
Then I added 1 CPU and now workers can start in less than 15s and sl-app can start normally.  I suppose this version is more heavy, but CPU is only used at startup, and is  0% in normal usage for selfhosting, then maybe is't a waste a resource only for a statup in 15s.

Anyway thanks for help.
